server/src/main/java/org/elasticsearch/cluster/coordination/JoinHelper.java

```diff
@@ -172,7 +176,57 @@ boolean isJoinPending() {
         return pendingOutgoingJoins.iterator().hasNext();
     }
 
-    void sendJoinRequest(DiscoveryNode destination, Optional<Join> optionalJoin) {
+    private static class FailedJoinAttempt {
+        private final DiscoveryNode destination;
+        private final JoinRequest joinRequest;
+        private final TransportException exception;
+        private final long timestamp;
+
+        FailedJoinAttempt(DiscoveryNode destination, JoinRequest joinRequest, TransportException exception) {
+            this.destination = destination;
+            this.joinRequest = joinRequest;
+            this.exception = exception;
+            this.timestamp = System.nanoTime();
+        }
+
+        void maybeLogNow() {
+            if (isSuspiciousTransportException(exception)) {
+                logger.info(() -> new ParameterizedMessage("failed to join {} with {}", destination, joinRequest), exception);
+            } else {
+                logger.debug(() -> new ParameterizedMessage("failed to join {} with {}", destination, joinRequest), exception);
+            }
+        }
+
+        boolean isSuspiciousTransportException(TransportException e) {
+            if (e instanceof RemoteTransportException) {
+                Throwable cause = e.getCause();
+                if (cause != null &&
```

Comment:
`instanceof` returns `false` on `null` so I think we don't need this first conjunct.

Reply:
It's fast-path check if it fails, we don't need to call `instanceof` 3 times. So I prefer to leave this check.